### ISSUE TITLE
Run z (test -O and -G)

### ISSUE THREAD

**Issue by jyn514** [oilshell/oil#422]
[`z`](https://github.com/rupa/z/) shows and navigates the most ['frecently'](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/Places/Frecency_algorithm) used directories.

Current crashes:
```
$ . ~/.bashrc
              precmd_functions[$(($#precmd_functions+1))]=_z_precmd
                                                    ^
/home/joshua/.local/lib/z:233: Expected first ) to end arith sub, got '+'
$ z
      [ -z "$_Z_OWNER" -a -f "$datafile" -a ! -O "$datafile" ] && return
      ^
/home/joshua/.local/lib/z:39: fatal: <Id_t BoolUnary_O 289> isn't implemented
```

**Comment by jyn514** [comment]
Hmm, that second crash only happens if "$datafile" exists, strange.

**Comment by jyn514** [comment]
The first crash seems to be because of static parsing, it's syntax specific to zsh:

```
 219   │ if type compctl >/dev/null 2>&1; then
 220   │     # zsh
 221   │     [ "$_Z_NO_PROMPT_COMMAND" ] || {
 222   │         # populate directory list, avoid clobbering any other precmds.
 223   │         if [ "$_Z_NO_RESOLVE_SYMLINKS" ]; then
 224   │             _z_precmd() {
 225   │                 (_z --add "${PWD:a}" &)
 226   │             }
 227   │         else
 228   │             _z_precmd() {
 229   │                 (_z --add "${PWD:A}" &)
 230   │             }
 231   │         fi
 232   │         [[ -n "${precmd_functions[(r)_z_precmd]}" ]] || {
 233   │             precmd_functions[$(($#precmd_functions+1))]=_z_precmd
 234   │         }
 235   │     }
 236   │     _z_zsh_tab_completion() {
 237   │         # tab completion
 238   │         local compl
 239   │         read -l compl
 240   │         reply=(${(f)"$(_z --complete "$compl")"})
 241   │     }
 242   │     compctl -U -K _z_zsh_tab_completion _z
```

**Comment by andychu** [comment]
Yeah I have run into the same issue with `git-completion` -- it mixed bash and zsh syntax.

The only solution is to factor `zsh` out into a file and `source` it, which should always work AFAIK. 

`-O` should be a one-liner.  I'm not sure what the system call that is, but I usually look in the dash source  when I have questions.

**Comment by jyn514** [comment]
The syscall is [`geteuid`](http://man7.org/linux/man-pages/man2/getuid.2.html).

```
read(0, "[ -O index.html ]\n", 8192)    = 18
stat("index.html", {st_mode=S_IFREG|0600, st_size=229, ...}) = 0
geteuid()                               = 1000
```

**Comment by andychu** [comment]
OK I added -O and -G.  I think that's the only thing we can do for now.

We could either maintain forks or try to upstream changes.  I did a lot of that with bash-completion.  But I got kinda sick of it and my strategy now is to make OSH more appealing in general, so other people (including the original authors) will be motivated to maintain OSH compatibility.

Factoring out zsh via `source` is a trivial change, so if the desire is there, it's a pretty easy path.

Thanks for the report -- more along these lines are great!
